### ISSUE TITLE
Interrupt in command substitution exits interactive bash

### ISSUE THREAD

**Issue by roberth** [NixOS/nixpkgs#101183]
**Describe the bug**

Bash exits instead of interrupting the child process.

**To Reproduce**

```shell
bash$ echo $(sleep 10)
^C
```

This exits the bash process, typically causing terminal emulator windows to close. I've tested this in Konsole, xfce4-terminal, xterm and gnome-terminal.
Interrupting `sleep 10` by itself works just fine.
Variations with command substitution in a different 'role' produce the same outcome. For example `$(sleep 10)` by itself or `a=$(sleep 10)`.

**Expected behavior**

The child process is interrupted and the shell doesn't exit.

**Additional context**

Tested with `bashInteractive_5` version `5.0-p18` on master. Also affects regular `bash`4.4.23(1) in 20.03, possibly other versions too.

Upstream confirms that the problem does not occur on Fedora 31 or macOS.
Here's the upstream issue https://savannah.gnu.org/support/?110335

**Notify maintainers**

@peti

**Metadata**
Please run `nix-shell -p nix-info --run "nix-info -m"` and paste the result.

Maintainer information:
```yaml
# a list of nixpkgs attributes affected by the problem
attribute:
  - bash
  - bash_5
  - bashInteractive
  - bashInteractive_5
# a list of nixos modules affected by the problem
module:
```

**Comment by JustinLovinger** [comment]
I can't reproduce the issue on NixOS 20.09 .

**Comment by roberth** [comment]
It seems to be triggered by direnv (https://github.com/direnv/direnv/issues/627) but it can only be solved in bash.

**Comment by roberth** [comment]
Fixed in bash 5.1 which has already become the default on NixOS by now.
